spot-contracts/contracts/vaults/RolloverVault.sol

```diff
@@ -594,4 +710,30 @@ contract RolloverVault is
             token.safeApprove(spender, type(uint256).max);
         }
     }
+
+    /// @dev Computes the underlying amount to be returned after deducting the melding fee.
+    ///      The meld fee is applied based on linear function based on the bond's time remaining to maturity wrt is duration.
+    function _computeMeldFee(IBondController bond, uint256 underlyingAmt) private view returns (uint256) {
+        uint256 maturityTimestampSec = bond.maturityDate();
+        uint256 bondDurationSec = maturityTimestampSec - bond.creationDate();
+        uint256 secondsToMaturity = (
+            maturityTimestampSec > block.timestamp ? maturityTimestampSec - block.timestamp : 0
+        );
+        uint256 appliedFeePerc = MAX_MELD_FEE_PERC.mulDiv(secondsToMaturity, bondDurationSec);
+        return underlyingAmt.mulDiv(appliedFeePerc, HUNDRED_PERC);
+    }
+
+    /// @dev Checks if at least one of the bond's tranches is part of the deployed set.
+    ///      Based on this condition we can infer that the bond is NOT malicious,
+    ///      and accept other sibling tranches from this bond.
+    function _isDeployedBond(IBondController bond, BondTranches memory bt) private view returns (bool) {
+        for (uint8 i = 0; i < bt.tranches.length; i++) {
+            // NOTE: We have to double check if the tranche in the deployed set
+            //       links back to the incoming bond, to ensure that the bond is not malicious.
+            if (_deployed.contains(address(bt.tranches[i])) && bt.tranches[i].bond() == address(bond)) {
```

Comment:
Could there be a malicious bond that shares one tranche with a real bond, but also has malicious tranches? Wouldn't we need to check every tranche?

Reply:
What we really want to check here is if any of the tranches which are already in the vault, point back to the incoming bond. We assume that the tranches in the system have been vetted and their references are truthful. 

A simple but gas inefficient will look like this:
```
bool whitelistedBond = false;
for(uint8 i=0; i<_deployed.length(); i++){
  whitelistedBond = whitelistedBond ||  ITranche(_deployed.at(i)).bond() == bond;
}
return whitelistedBond;
```
 
Alternatively, we can iterate through the incoming bond's tranches to check if one of them are already in the vault. This should be more gas efficient as the tranche_len will be 2/3 and the vault_len will be much higher (and grows as a multiple of the tranche_len).

This still doesn't guard against malicious bonds. As a malicious bond fake the references to just have one "real" tranche (which is in the system). The 2nd check guards against this, the tranche in the system should link back to the actual whitelisted parent (from the spot issuer). We check if the incoming bond, matches the parent bond of the matched tranche in the system.

Reply:
Another way to do this is to reverse traverse Spot's Issuer's activeBondIDXs or issuedAt
- the bond will be active if not matured, but the bondIssuer's `issued` list is ordered by issued date, so theoretically won't have to check more than 4 indexes
- the bond will have to have been issued by spot's bond issuer

The only issue is that the BondIssuer can be updated.. but maybe we can add an 'old' issuers list? this would only ever be needed for the few bonds issued right before an update

Reply:
Right, we don't really keep track of the list of bonds in perp anywhere. 

I think we should do an efficiency pass down the road. We can clean up the code a lot if we assume that a Bond has only 2 tranches. (and the list of tranches in perp will correspond to the set of bonds in the system)